src/wp-includes/class-wp-speculation-rules.php

```diff
@@ -0,0 +1,218 @@
+			);
+			return false;
+		}
+		if ( isset( $rule['source'] ) ) {
+			if ( ! wp_is_valid_speculation_rules_source( $rule['source'] ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: %s: invalid source value */
+						__( 'The value "%s" is not a valid source for a speculation rule.' ),
+						esc_html( $rule['source'] )
+					),
+					'6.8.0'
+				);
+				return false;
+			}
+
+			if ( 'list' === $rule['source'] && isset( $rule['where'] ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: 1: source value, 2: forbidden key */
+						__( 'A speculation rule of source "%1$s" must not include a "%2$s" key.' ),
+						'list',
+						'where'
+					),
+					'6.8.0'
+				);
+				return false;
+			}
+
+			if ( 'document' === $rule['source'] && isset( $rule['urls'] ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					sprintf(
+						/* translators: 1: source value, 2: forbidden key */
+						__( 'A speculation rule of source "%1$s" must not include a "%2$s" key.' ),
+						'document',
+						'urls'
+					),
+					'6.8.0'
+				);
+				return false;
+			}
+		}
+
+		// If there is an 'eagerness' key specified, make sure it's valid.
+		if ( isset( $rule['eagerness'] ) && ! wp_is_valid_speculation_rules_eagerness( $rule['eagerness'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				sprintf(
+					/* translators: %s: invalid eagerness value */
+					__( 'The value "%s" is not a valid eagerness for a speculation rule.' ),
+					esc_html( $rule['eagerness'] )
+				),
+				'6.8.0'
+			);
+			return false;
+		}
+
+		if ( ! isset( $this->rules_by_mode[ $mode ] ) ) {
+			$this->rules_by_mode[ $mode ] = array();
+		}
+
+		$this->rules_by_mode[ $mode ][ $id ] = $rule;
+		return true;
+	}
+
+	/**
+	 * Checks whether a speculation rule for the given mode and ID already exists.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @param string $mode Speculative loading mode. Either 'prefetch' or 'prerender'.
+	 * @param string $id   Unique string identifier for the speculation rule.
+	 * @return bool True if the rule already exists, false otherwise.
+	 */
+	public function has_rule( string $mode, string $id ): bool {
+		return isset( $this->rules_by_mode[ $mode ][ $id ] );
+	}
+
+	/**
+	 * Returns the associative array representation of the speculation rules data.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @return array<string, array<string, mixed>> Speculation rules data.
+	 */
+	public function to_array(): array {
+		return array_filter( $this->rules_by_mode );
+	}
+
+	/**
+	 * Returns the speculation rules data ready to be JSON-encoded.
+	 *
+	 * @since 6.8.0
+	 *
+	 * @return array<string, array<string, mixed>> Speculation rules data.
+	 */
+	#[ReturnTypeWillChange]
+	public function jsonSerialize() {
+		// For the JSON output, we need to strip the IDs, which are not relevant for the Speculation Rules API.
+		return array_map(
+			static function ( array $rules ) {
+				return array_values( $rules );
+			},
+			$this->to_array()
```

Comment:
```suggestion
			array_filter( $this->rules_by_mode )
```

Reply:
If we remove `to_array()`, this makes sense, otherwise I don't see why this shouldn't use `to_array()`.

Reply:
Updated as part of https://github.com/WordPress/wordpress-develop/pull/7860/commits/d4c382d23c9824306a7b39028a3a109c394c04e4